File: web/src/i18n/content-languages.ts
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
// Content languages offered for public content (blog, legal pages, SEO,
// model descriptions). Mirrors common/content_languages.go on the backend;
// keep both in sync. English is the required base language, everything else
// is optional and falls back to English (or a machine translation).

export type ContentLanguage = {
  code: string
  name: string
  native: string
  flag: string
}

export const DEFAULT_CONTENT_LANGUAGE = 'en'

export const CONTENT_LANGUAGES: ContentLanguage[] = [
  { code: 'en', name: 'English', native: 'English', flag: '🇬🇧' },
  {
    code: 'zh-CN',
    name: 'Chinese (Simplified)',
    native: '简体中文',
    flag: '🇨🇳',
  },
  {
    code: 'zh-TW',
    name: 'Chinese (Traditional)',
    native: '繁體中文',
    flag: '🇹🇼',
  },
  { code: 'tr', name: 'Turkish', native: 'Türkçe', flag: '🇹🇷' },
  { code: 'fr', name: 'French', native: 'Français', flag: '🇫🇷' },
  { code: 'de', name: 'German', native: 'Deutsch', flag: '🇩🇪' },
  { code: 'es', name: 'Spanish', native: 'Español', flag: '🇪🇸' },
  { code: 'pt', name: 'Portuguese', native: 'Português', flag: '🇵🇹' },
  { code: 'ru', name: 'Russian', native: 'Русский', flag: '🇷🇺' },
  { code: 'ja', name: 'Japanese', native: '日本語', flag: '🇯🇵' },
  { code: 'ko', name: 'Korean', native: '한국어', flag: '🇰🇷' },
  { code: 'ar', name: 'Arabic', native: 'العربية', flag: '🇸🇦' },
  { code: 'hi', name: 'Hindi', native: 'हिन्दी', flag: '🇮🇳' },
  { code: 'id', name: 'Indonesian', native: 'Bahasa Indonesia', flag: '🇮🇩' },
  { code: 'vi', name: 'Vietnamese', native: 'Tiếng Việt', flag: '🇻🇳' },
  { code: 'th', name: 'Thai', native: 'ไทย', flag: '🇹🇭' },
  { code: 'it', name: 'Italian', native: 'Italiano', flag: '🇮🇹' },
  { code: 'nl', name: 'Dutch', native: 'Nederlands', flag: '🇳🇱' },
  { code: 'pl', name: 'Polish', native: 'Polski', flag: '🇵🇱' },
  { code: 'uk', name: 'Ukrainian', native: 'Українська', flag: '🇺🇦' },
  { code: 'cs', name: 'Czech', native: 'Čeština', flag: '🇨🇿' },
  { code: 'sk', name: 'Slovak', native: 'Slovenčina', flag: '🇸🇰' },
  { code: 'sl', name: 'Slovenian', native: 'Slovenščina', flag: '🇸🇮' },
  { code: 'hr', name: 'Croatian', native: 'Hrvatski', flag: '🇭🇷' },
  { code: 'sr', name: 'Serbian', native: 'Српски', flag: '🇷🇸' },
  { code: 'bs', name: 'Bosnian', native: 'Bosanski', flag: '🇧🇦' },
  { code: 'bg', name: 'Bulgarian', native: 'Български', flag: '🇧🇬' },
  { code: 'mk', name: 'Macedonian', native: 'Македонски', flag: '🇲🇰' },
  { code: 'el', name: 'Greek', native: 'Ελληνικά', flag: '🇬🇷' },
  { code: 'ro', name: 'Romanian', native: 'Română', flag: '🇷🇴' },
  { code: 'hu', name: 'Hungarian', native: 'Magyar', flag: '🇭🇺' },
  { code: 'da', name: 'Danish', native: 'Dansk', flag: '🇩🇰' },
  { code: 'sv', name: 'Swedish', native: 'Svenska', flag: '🇸🇪' },
  { code: 'no', name: 'Norwegian', native: 'Norsk', flag: '🇳🇴' },
  { code: 'fi', name: 'Finnish', native: 'Suomi', flag: '🇫🇮' },
  { code: 'et', name: 'Estonian', native: 'Eesti', flag: '🇪🇪' },
  { code: 'lv', name: 'Latvian', native: 'Latviešu', flag: '🇱🇻' },
  { code: 'lt', name: 'Lithuanian', native: 'Lietuvių', flag: '🇱🇹' },
  { code: 'is', name: 'Icelandic', native: 'Íslenska', flag: '🇮🇸' },
  { code: 'ga', name: 'Irish', native: 'Gaeilge', flag: '🇮🇪' },
  { code: 'mt', name: 'Maltese', native: 'Malti', flag: '🇲🇹' },
  { code: 'sq', name: 'Albanian', native: 'Shqip', flag: '🇦🇱' },
  { code: 'ca', name: 'Catalan', native: 'Català', flag: '🇪🇸' },
  { code: 'eu', name: 'Basque', native: 'Euskara', flag: '🇪🇸' },
  { code: 'gl', name: 'Galician', native: 'Galego', flag: '🇪🇸' },
  { code: 'he', name: 'Hebrew', native: 'עברית', flag: '🇮🇱' },
  { code: 'fa', name: 'Persian', native: 'فارسی', flag: '🇮🇷' },
  { code: 'ur', name: 'Urdu', native: 'اردو', flag: '🇵🇰' },
  { code: 'bn', name: 'Bengali', native: 'বাংলা', flag: '🇧🇩' },
  { code: 'ta', name: 'Tamil', native: 'தமிழ்', flag: '🇮🇳' },
  { code: 'te', name: 'Telugu', native: 'తెలుగు', flag: '🇮🇳' },
  { code: 'mr', name: 'Marathi', native: 'मराठी', flag: '🇮🇳' },
  { code: 'gu', name: 'Gujarati', native: 'ગુજરાતી', flag: '🇮🇳' },
  { code: 'kn', name: 'Kannada', native: 'ಕನ್ನಡ', flag: '🇮🇳' },
  { code: 'ml', name: 'Malayalam', native: 'മലയാളം', flag: '🇮🇳' },
  { code: 'pa', name: 'Punjabi', native: 'ਪੰਜਾਬੀ', flag: '🇮🇳' },
  { code: 'sw', name: 'Swahili', native: 'Kiswahili', flag: '🇰🇪' },
  { code: 'am', name: 'Amharic', native: 'አማርኛ', flag: '🇪🇹' },
  { code: 'ms', name: 'Malay', native: 'Bahasa Melayu', flag: '🇲🇾' },
  { code: 'tl', name: 'Filipino', native: 'Filipino', flag: '🇵🇭' },
  { code: 'my', name: 'Burmese', native: 'မြန်မာ', flag: '🇲🇲' },
  { code: 'km', name: 'Khmer', native: 'ខ្មែរ', flag: '🇰🇭' },
  { code: 'lo', name: 'Lao', native: 'ລາວ', flag: '🇱🇦' },
  { code: 'ka', name: 'Georgian', native: 'ქართული', flag: '🇬🇪' },
  { code: 'hy', name: 'Armenian', native: 'Հայերեն', flag: '🇦🇲' },
  { code: 'az', name: 'Azerbaijani', native: 'Azərbaycan', flag: '🇦🇿' },
  { code: 'kk', name: 'Kazakh', native: 'Қазақша', flag: '🇰🇿' },
  { code: 'uz', name: 'Uzbek', native: 'Oʻzbek', flag: '🇺🇿' },
  { code: 'mn', name: 'Mongolian', native: 'Монгол', flag: '🇲🇳' },
  { code: 'ne', name: 'Nepali', native: 'नेपाली', flag: '🇳🇵' },
  { code: 'si', name: 'Sinhala', native: 'සිංහල', flag: '🇱🇰' },
  { code: 'af', name: 'Afrikaans', native: 'Afrikaans', flag: '🇿🇦' },
]

const contentLanguageByCode = new Map(
  CONTENT_LANGUAGES.map((lang) => [lang.code.toLowerCase(), lang])
)

/** Canonicalize a language tag ("TR", "zh_cn") to a content language code. */
export function normalizeContentLanguage(value?: string | null): string {
  if (!value) return ''
  const normalized = value.trim().replaceAll('_', '-').toLowerCase()
  return contentLanguageByCode.get(normalized)?.code ?? ''
}

/** Map an i18next interface code (zhCN/zhTW) to a content language code. */
export function interfaceToContentLanguage(code?: string | null): string {
  if (code === 'zhCN') return 'zh-CN'
  if (code === 'zhTW') return 'zh-TW'
  return normalizeContentLanguage(code) || DEFAULT_CONTENT_LANGUAGE
}

/** Map a content language code back to the i18next interface code. */
export function contentToInterfaceLanguage(code?: string | null): string {
  const normalized = normalizeContentLanguage(code)
  if (normalized === 'zh-CN') return 'zhCN'
  if (normalized === 'zh-TW') return 'zhTW'
  return normalized || DEFAULT_CONTENT_LANGUAGE
}

/** Extract a language prefix from a URL path: "/tr/blog" -> { lang: "tr", strippedPath: "/blog" }. */
export function extractLanguagePrefix(pathname: string): {
  lang: string
  strippedPath: string
} {
  const match = pathname.match(/^\/([A-Za-z-]{2,5})(\/|$)/)
  if (!match) return { lang: '', strippedPath: pathname }
  const lang = normalizeContentLanguage(match[1])
  if (!lang) return { lang: '', strippedPath: pathname }
  const strippedPath =
    match[2] === '/' ? pathname.slice(match[1].length + 1) : '/'
  return { lang, strippedPath: strippedPath || '/' }
}

File: common/content_languages.go
package common

import "strings"

// ContentLanguage describes one language the site's public content (blog,
// legal pages, SEO metadata, model descriptions) can be served in. The base
// language is English; every other language is optional and falls back to
// English (or an on-the-fly machine translation) when missing.
type ContentLanguage struct {
	Code   string `json:"code"`   // BCP-47 tag, e.g. "tr", "zh-CN"
	Name   string `json:"name"`   // English name
	Native string `json:"native"` // native name shown in the language picker
	Flag   string `json:"flag"`
}

const DefaultContentLanguage = "en"

// ContentLanguages is the canonical list of languages offered for public
// content. The frontend keeps a mirror of this list for boot-time URL prefix
// detection; keep both in sync.
var ContentLanguages = []ContentLanguage{
	{Code: "en", Name: "English", Native: "English", Flag: "🇬🇧"},
	{Code: "zh-CN", Name: "Chinese (Simplified)", Native: "简体中文", Flag: "🇨🇳"},
	{Code: "zh-TW", Name: "Chinese (Traditional)", Native: "繁體中文", Flag: "🇹🇼"},
	{Code: "tr", Name: "Turkish", Native: "Türkçe", Flag: "🇹🇷"},
	{Code: "fr", Name: "French", Native: "Français", Flag: "🇫🇷"},
	{Code: "de", Name: "German", Native: "Deutsch", Flag: "🇩🇪"},
	{Code: "es", Name: "Spanish", Native: "Español", Flag: "🇪🇸"},
	{Code: "pt", Name: "Portuguese", Native: "Português", Flag: "🇵🇹"},
	{Code: "ru", Name: "Russian", Native: "Русский", Flag: "🇷🇺"},
	{Code: "ja", Name: "Japanese", Native: "日本語", Flag: "🇯🇵"},
	{Code: "ko", Name: "Korean", Native: "한국어", Flag: "🇰🇷"},
	{Code: "ar", Name: "Arabic", Native: "العربية", Flag: "🇸🇦"},
	{Code: "hi", Name: "Hindi", Native: "हिन्दी", Flag: "🇮🇳"},
	{Code: "id", Name: "Indonesian", Native: "Bahasa Indonesia", Flag: "🇮🇩"},
	{Code: "vi", Name: "Vietnamese", Native: "Tiếng Việt", Flag: "🇻🇳"},
	{Code: "th", Name: "Thai", Native: "ไทย", Flag: "🇹🇭"},
	{Code: "it", Name: "Italian", Native: "Italiano", Flag: "🇮🇹"},
	{Code: "nl", Name: "Dutch", Native: "Nederlands", Flag: "🇳🇱"},
	{Code: "pl", Name: "Polish", Native: "Polski", Flag: "🇵🇱"},
	{Code: "uk", Name: "Ukrainian", Native: "Українська", Flag: "🇺🇦"},
	{Code: "cs", Name: "Czech", Native: "Čeština", Flag: "🇨🇿"},
	{Code: "sk", Name: "Slovak", Native: "Slovenčina", Flag: "🇸🇰"},
	{Code: "sl", Name: "Slovenian", Native: "Slovenščina", Flag: "🇸🇮"},
	{Code: "hr", Name: "Croatian", Native: "Hrvatski", Flag: "🇭🇷"},
	{Code: "sr", Name: "Serbian", Native: "Српски", Flag: "🇷🇸"},
	{Code: "bs", Name: "Bosnian", Native: "Bosanski", Flag: "🇧🇦"},
	{Code: "bg", Name: "Bulgarian", Native: "Български", Flag: "🇧🇬"},
	{Code: "mk", Name: "Macedonian", Native: "Македонски", Flag: "🇲🇰"},
	{Code: "el", Name: "Greek", Native: "Ελληνικά", Flag: "🇬🇷"},
	{Code: "ro", Name: "Romanian", Native: "Română", Flag: "🇷🇴"},
	{Code: "hu", Name: "Hungarian", Native: "Magyar", Flag: "🇭🇺"},
	{Code: "da", Name: "Danish", Native: "Dansk", Flag: "🇩🇰"},
	{Code: "sv", Name: "Swedish", Native: "Svenska", Flag: "🇸🇪"},
	{Code: "no", Name: "Norwegian", Native: "Norsk", Flag: "🇳🇴"},
	{Code: "fi", Name: "Finnish", Native: "Suomi", Flag: "🇫🇮"},
	{Code: "et", Name: "Estonian", Native: "Eesti", Flag: "🇪🇪"},
	{Code: "lv", Name: "Latvian", Native: "Latviešu", Flag: "🇱🇻"},
	{Code: "lt", Name: "Lithuanian", Native: "Lietuvių", Flag: "🇱🇹"},
	{Code: "is", Name: "Icelandic", Native: "Íslenska", Flag: "🇮🇸"},
	{Code: "ga", Name: "Irish", Native: "Gaeilge", Flag: "🇮🇪"},
	{Code: "mt", Name: "Maltese", Native: "Malti", Flag: "🇲🇹"},
	{Code: "sq", Name: "Albanian", Native: "Shqip", Flag: "🇦🇱"},
	{Code: "ca", Name: "Catalan", Native: "Català", Flag: "🇪🇸"},
	{Code: "eu", Name: "Basque", Native: "Euskara", Flag: "🇪🇸"},
	{Code: "gl", Name: "Galician", Native: "Galego", Flag: "🇪🇸"},
	{Code: "he", Name: "Hebrew", Native: "עברית", Flag: "🇮🇱"},
	{Code: "fa", Name: "Persian", Native: "فارسی", Flag: "🇮🇷"},
	{Code: "ur", Name: "Urdu", Native: "اردو", Flag: "🇵🇰"},
	{Code: "bn", Name: "Bengali", Native: "বাংলা", Flag: "🇧🇩"},
	{Code: "ta", Name: "Tamil", Native: "தமிழ்", Flag: "🇮🇳"},
	{Code: "te", Name: "Telugu", Native: "తెలుగు", Flag: "🇮🇳"},
	{Code: "mr", Name: "Marathi", Native: "मराठी", Flag: "🇮🇳"},
	{Code: "gu", Name: "Gujarati", Native: "ગુજરાતી", Flag: "🇮🇳"},
	{Code: "kn", Name: "Kannada", Native: "ಕನ್ನಡ", Flag: "🇮🇳"},
	{Code: "ml", Name: "Malayalam", Native: "മലയാളം", Flag: "🇮🇳"},
	{Code: "pa", Name: "Punjabi", Native: "ਪੰਜਾਬੀ", Flag: "🇮🇳"},
	{Code: "sw", Name: "Swahili", Native: "Kiswahili", Flag: "🇰🇪"},
	{Code: "am", Name: "Amharic", Native: "አማርኛ", Flag: "🇪🇹"},
	{Code: "ms", Name: "Malay", Native: "Bahasa Melayu", Flag: "🇲🇾"},
	{Code: "tl", Name: "Filipino", Native: "Filipino", Flag: "🇵🇭"},
	{Code: "my", Name: "Burmese", Native: "မြန်မာ", Flag: "🇲🇲"},
	{Code: "km", Name: "Khmer", Native: "ខ្មែរ", Flag: "🇰🇭"},
	{Code: "lo", Name: "Lao", Native: "ລາວ", Flag: "🇱🇦"},
	{Code: "ka", Name: "Georgian", Native: "ქართული", Flag: "🇬🇪"},
	{Code: "hy", Name: "Armenian", Native: "Հայերեն", Flag: "🇦🇲"},
	{Code: "az", Name: "Azerbaijani", Native: "Azərbaycan", Flag: "🇦🇿"},
	{Code: "kk", Name: "Kazakh", Native: "Қазақша", Flag: "🇰🇿"},
	{Code: "uz", Name: "Uzbek", Native: "Oʻzbek", Flag: "🇺🇿"},
	{Code: "mn", Name: "Mongolian", Native: "Монгол", Flag: "🇲🇳"},
	{Code: "ne", Name: "Nepali", Native: "नेपाली", Flag: "🇳🇵"},
	{Code: "si", Name: "Sinhala", Native: "සිංහල", Flag: "🇱🇰"},
	{Code: "af", Name: "Afrikaans", Native: "Afrikaans", Flag: "🇿🇦"},
}

var contentLanguageByCode = func() map[string]ContentLanguage {
	m := make(map[string]ContentLanguage, len(ContentLanguages))
	for _, lang := range ContentLanguages {
		m[strings.ToLower(lang.Code)] = lang
	}
	return m
}()

// NormalizeContentLanguage canonicalizes a language tag ("TR", "zh_cn",
// "zh-tw" -> "tr", "zh-CN", "zh-TW"). Returns "" for unknown codes.
func NormalizeContentLanguage(code string) string {
	lang, ok := LookupContentLanguage(code)
	if !ok {
		return ""
	}
	return lang.Code
}

// LookupContentLanguage resolves a user-supplied tag (any casing, "_" or "-"
// separators) to a supported content language.
func LookupContentLanguage(code string) (ContentLanguage, bool) {
	normalized := strings.ToLower(strings.ReplaceAll(strings.TrimSpace(code), "_", "-"))
	if normalized == "" {
		return ContentLanguage{}, false
	}
	lang, ok := contentLanguageByCode[normalized]
	return lang, ok
}

// IsContentLanguage reports whether the code is a supported content language.
func IsContentLanguage(code string) bool {
	_, ok := LookupContentLanguage(code)
	return ok
}
